src/device_rx.c

```diff
@@ -0,0 +1,581 @@
+            &iter, NULL, &interface_name_size, NULL, &path_size);
+        if (ares != ASTARTE_RESULT_OK) {
+            ASTARTE_LOG_ERR("Properties iterator get error: %s", astarte_result_to_name(ares));
+            goto end;
+        }
+
+        // Allocate space for the name and path
+        interface_name = calloc(interface_name_size, sizeof(char));
+        path = calloc(path_size, sizeof(char));
+        if (!interface_name || !path) {
+            ASTARTE_LOG_ERR("Out of memory %s: %d", __FILE__, __LINE__);
+            goto end;
+        }
+
+        ares = device_caching_property_iterator_get(
+            &iter, interface_name, &interface_name_size, path, &path_size);
+        if (ares != ASTARTE_RESULT_OK) {
+            ASTARTE_LOG_ERR("Properties iterator get error: %s", astarte_result_to_name(ares));
+            goto end;
+        }
+
+        // Purge the property if not in the allow list
+        purge_server_property(introspection, interface_name, path, allow_list);
+
+        free(interface_name);
+        interface_name = NULL;
+        free(path);
+        path = NULL;
+
+        ares = device_caching_property_iterator_next(&iter);
+        if ((ares != ASTARTE_RESULT_OK) && (ares != ASTARTE_RESULT_NOT_FOUND)) {
+            ASTARTE_LOG_ERR("Iterator next error: %s", astarte_result_to_name(ares));
+            goto end;
+        }
+    }
+
+end:
+    device_caching_property_iterator_terminate(&iter);
+    free(interface_name);
+    free(path);
+}
+
+static void purge_server_property(
+    introspection_t *introspection, char *interface_name, char *path, dlist_t allow_list)
+{
+    astarte_result_t ares = ASTARTE_RESULT_OK;
+    char *property = NULL;
+
+    const astarte_interface_t *interface = introspection_get(introspection, interface_name);
+    if (!interface) {
+        ASTARTE_LOG_DBG("Purging property from unknown interface: '%s%s'", interface_name, path);
+        ares = device_caching_property_delete(interface_name, path);
+        if ((ares != ASTARTE_RESULT_OK) && (ares != ASTARTE_RESULT_NOT_FOUND)) {
+            if (ares != ASTARTE_RESULT_OK) {
+                ASTARTE_LOG_ERR(
+                    "Failed deleting the cached property: %s", astarte_result_to_name(ares));
+            }
+        }
+        goto end;
+    }
+
+    if (interface->ownership != ASTARTE_INTERFACE_OWNERSHIP_SERVER) {
+        goto end;
+    }
+
+    // Concatenate the interface_name and path
+    property = calloc(strlen(interface_name) + strlen(path) + 1, sizeof(char));
+    if (!property) {
+        ASTARTE_LOG_ERR("Out of memory %s: %d", __FILE__, __LINE__);
+        goto end;
+    }
+    int snprintf_rc
+        = snprintf(property, strlen(interface_name) + strlen(path), "%s%s", interface_name, path);
+    if (snprintf_rc != strlen(interface_name) + strlen(path)) {
+        ASTARTE_LOG_ERR("Error encoding interface name '%s' and path '%s' in a single string.",
+            interface_name, path);
+        goto end;
+    }
```

Comment:
What do you think about defining the property_len instead of calculating it every time?
```suggestion
    size_t property_len = strlen(interface_name) + strlen(path) ;
    property = calloc(property_len + 1, sizeof(char));
    if (!property) {
        ASTARTE_LOG_ERR("Out of memory %s: %d", __FILE__, __LINE__);
        goto end;
    }
    int snprintf_rc
        = snprintf(property, property_len, "%s%s", interface_name, path);
    if (snprintf_rc != property_len ) {
        ASTARTE_LOG_ERR("Error encoding interface name '%s' and path '%s' in a single string.",
            interface_name, path);
        goto end;
    }
```